Unable to parse class with generic parameter using JavaParserTypeSolver

Hello guys.
First of all, thanks for this amazing tool!

When I was working with it, i noticed that sometimes the _JavaParserTypeSolver_ was unable to resolve some of the Classes.
When diving deeper in what was going on, noticed that all of them were generic types, like the class C in the example below.

```java
package com.example;

public class A<X> {

    private class B {}

    private class C<Z> extends A<Z>.B {}

}
```

During debuging found out the issue was related with the **JavaParserTypeSolver** that tried to look for a file that doesn't exist.
In the given example the file would look for "/A\<Z\>/B.java" instead of Class B inside of "A.java".


#### Expected Behavior

When resolving a type that can't be found by a type solver, it should throw a _UnsolvedSymbolException_ (Based on [documentation](https://javadoc.io/doc/com.github.javaparser/javaparser-symbol-solver-model/latest/com/github/javaparser/symbolsolver/model/resolution/TypeSolver.html#solveType-java.lang.String-)).

#### Current Behavior

When resolving the class an _UncheckedExecutionException_ is thrown.
```
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.nio.file.InvalidPathException: Illegal char <<> at index 65: C:\Users\User\AppData\Local\Temp\tempProject6055408844541354802/A<Z>/B.java
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4848)
	at com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver.tryToSolveType(JavaParserTypeSolver.java:202)
	at com.github.javaparser.symbolsolver.model.resolution.TypeSolver.solveType(TypeSolver.java:66)
	at pt.forever.university.analyser.TestBug.test0(TestBug.java:788)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	... 17 more
Caused by: java.nio.file.InvalidPathException: Illegal char <<> at index 65: C:\Users\User\AppData\Local\Temp\tempProject6055408844541354802/A<Z>/B.java
	at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at sun.nio.fs.WindowsPath.parse(WindowsPath.java:94)
	at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:255)
	at java.nio.file.Paths.get(Paths.java:84)
	at com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver.tryToSolveTypeUncached(JavaParserTypeSolver.java:234)
	at com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver.lambda$tryToSolveType$4(JavaParserTypeSolver.java:203)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4853)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 31 more
```

#### Test for issue

The following test can recreate the issue.
The test is composed of the following steps:

1. Create a path with the root sources (may be empty);
2. Create a new JavaParserTypeSolver and provide the root sources as parameter;
3. Use the created type solver to resolve a type with inner class and a outer class where the outer class has a generic parameter.

Test 1:
```java
@Test(expected = UnsolvedSymbolException.class)
public void testUsingParameterizedClass() throws IOException {

    Path temporaryProject = Files.createTempDirectory("temporaryProject");
    new JavaParserTypeSolver(temporaryProject)
            .solveType("A<Z>");
}
```

Test 2:
```java
@Test(expected = UnsolvedSymbolException.class)
public void testUsingParameterizedOuterClass() throws IOException {

    Path temporaryProject = Files.createTempDirectory("temporaryProject");
    new JavaParserTypeSolver(temporaryProject)
            .solveType("A<Z>.B");
}
```

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java
@@ -21,16 +21,9 @@
 
 package com.github.javaparser.symbolsolver.resolution.typesolvers;
 
-import com.github.javaparser.JavaParser;
-import com.github.javaparser.ParserConfiguration;
-import com.github.javaparser.ast.CompilationUnit;
-import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
-import com.github.javaparser.symbolsolver.javaparser.Navigator;
-import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
-import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
-import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
-import com.google.common.cache.Cache;
-import com.google.common.cache.CacheBuilder;
+import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
+import static com.github.javaparser.ParserConfiguration.LanguageLevel.BLEEDING_EDGE;
+import static com.github.javaparser.Providers.provider;
 
 import java.io.File;
 import java.io.FileNotFoundException;
@@ -44,9 +37,17 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
-import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
-import static com.github.javaparser.ParserConfiguration.LanguageLevel.BLEEDING_EDGE;
-import static com.github.javaparser.Providers.provider;
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import com.github.javaparser.symbolsolver.javaparser.Navigator;
+import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
+import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.utils.FileUtils;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 
 /**
  * Defines a directory containing source code that should be used for solving symbols.
@@ -217,7 +218,7 @@ public class JavaParserTypeSolver implements TypeSolver {
         for (int i = nameElements.length; i > 0; i--) {
             StringBuilder filePath = new StringBuilder(srcDir.toAbsolutePath().toString());
             for (int j = 0; j < i; j++) {
-                filePath.append("/")
+                filePath.append(File.separator)
                         .append(nameElements[j]);
             }
             filePath.append(".java");
@@ -230,26 +231,34 @@ public class JavaParserTypeSolver implements TypeSolver {
                 typeName.append(nameElements[j]);
             }
 
+            String dirToParse = null;
             // As an optimization we first try to look in the canonical position where we expect to find the file
-            Path srcFile = Paths.get(filePath.toString());
-            {
+            if (FileUtils.isValidPath(filePath.toString())) {
+                Path srcFile = Paths.get(filePath.toString());
                 Optional<CompilationUnit> compilationUnit = parse(srcFile);
                 if (compilationUnit.isPresent()) {
-                    Optional<com.github.javaparser.ast.body.TypeDeclaration<?>> astTypeDeclaration = Navigator.findType(compilationUnit.get(), typeName.toString());
+                    Optional<com.github.javaparser.ast.body.TypeDeclaration<?>> astTypeDeclaration = Navigator
+                            .findType(compilationUnit.get(), typeName.toString());
                     if (astTypeDeclaration.isPresent()) {
-                        return SymbolReference.solved(JavaParserFacade.get(this).getTypeDeclaration(astTypeDeclaration.get()));
+                        return SymbolReference
+                                .solved(JavaParserFacade.get(this).getTypeDeclaration(astTypeDeclaration.get()));
                     }
                 }
+                dirToParse = srcFile.getParent().normalize().toString();
+            } else {
+                dirToParse = FileUtils.getParentPath(filePath.toString());
             }
 
             // If this is not possible we parse all files
             // We try just in the same package, for classes defined in a file not named as the class itself
-            {
-                List<CompilationUnit> compilationUnits = parseDirectory(srcFile.getParent());
+            if (FileUtils.isValidPath(dirToParse)) {
+                List<CompilationUnit> compilationUnits = parseDirectory(Paths.get(dirToParse));
                 for (CompilationUnit compilationUnit : compilationUnits) {
-                    Optional<com.github.javaparser.ast.body.TypeDeclaration<?>> astTypeDeclaration = Navigator.findType(compilationUnit, typeName.toString());
+                    Optional<com.github.javaparser.ast.body.TypeDeclaration<?>> astTypeDeclaration = Navigator
+                            .findType(compilationUnit, typeName.toString());
                     if (astTypeDeclaration.isPresent()) {
-                        return SymbolReference.solved(JavaParserFacade.get(this).getTypeDeclaration(astTypeDeclaration.get()));
+                        return SymbolReference
+                                .solved(JavaParserFacade.get(this).getTypeDeclaration(astTypeDeclaration.get()));
                     }
                 }
             }

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/utils/FileUtils.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/utils/FileUtils.java
@@ -1,0 +1,26 @@
+package com.github.javaparser.symbolsolver.utils;
+
+import java.io.File;
+
+import com.github.javaparser.utils.Utils;
+
+public class FileUtils {
+    
+    /*
+     * returns true if the filename exists otherwise return false
+     */
+    public static boolean isValidPath(String filename) {
+        File file = new File(filename);
+        return file.exists();
+    }
+    
+    /*
+     * returns the parent path from the filename as string
+     */
+    public static String getParentPath(String filename) {
+        Utils.assertNotNull(filename);
+        int lastIndex = filename.lastIndexOf(File.separator);
+        return filename.substring(0, lastIndex);
+    }
+
+}

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolverTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolverTest.java
@@ -21,6 +21,19 @@
 
 package com.github.javaparser.symbolsolver.resolution.typesolvers;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
 import com.github.javaparser.ast.Node;
 import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
 import com.github.javaparser.symbolsolver.AbstractSymbolResolutionTest;
@@ -28,15 +41,6 @@ import com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParse
 import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
 import com.github.javaparser.symbolsolver.utils.LeanParserConfiguration;
 import com.github.javaparser.utils.CodeGenerationUtils;
-import org.junit.jupiter.api.Disabled;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.io.TempDir;
-
-import java.io.File;
-import java.io.IOException;
-import java.nio.file.Path;
-
-import static org.junit.jupiter.api.Assertions.*;
 
 class JavaParserTypeSolverTest extends AbstractSymbolResolutionTest {
 
@@ -126,6 +130,16 @@ class JavaParserTypeSolverTest extends AbstractSymbolResolutionTest {
         assertTrue(x.isSolved());
         assertNotNull(x.getCorrespondingDeclaration());
         assertTrue(x.getCorrespondingDeclaration().isInterface());
+    }
+    
+    @Test
+    public void givenJavaParserTypeSolver_tryToSolveAnUnexpectedSourceFileName_expectSuccess() {
+        Path src = adaptPath("src/test/test_sourcecode");
+        JavaParserTypeSolver typeSolver = new JavaParserTypeSolver(src);
+
+        SymbolReference<ResolvedReferenceTypeDeclaration> x = typeSolver.tryToSolveType("A<>");
+
+        assertFalse(x.isSolved());
     }
 
 }

